### PR TITLE
Added the implementation for login and signup endpoints.

### PR DESCRIPTION
**Pull request by santhoshMalay**
Signup example: - 
{
  "display_name": "msk1",
  "email": "msk1@gmail.com",
  "first_name": "mskFirst",
  "last_name": "mskLast",
  "password": "123456"
}

login example: -
{
  "email": "msk1@gmail.com",
  "password": "123456",
  "persistent": true
}

### CODE REVIEW

File: internal/service/users.go [hunk 1]
```diff
@@ -40,6 +45,49 @@ func (u *usersService) UpdateUserInfo(ctx context.Context, id string, input *Upd
...
+func (u *usersService) Signup(ctx context.Context, input *SignupUserInput) error {
+	user, err := u.repo.GetByEmail(ctx, input.Email)
+	if err != nil && err != repository.ErrNotFound {
+		return err
+	}
+
+	if user != nil && user.Email == input.Email {
+		return ErrUserAlreadyExist
+	}
+
```
Comment (zhuravlev-pe) [inline internal/service/users.go:68]: For the second argument I suggest using bcrypt.DefaultCost (10) at least. This parameter has an exponential effect, so the difference is between MinCost (4) and the default one is huge.

This is a training project, and actual security means little. On the other hand, we are here to learn some proper practices, so we should not use cryptography parameters which are highly insecure.

The thing is, the purpose of bcrypt is the protection against brute-forcing password hashes (e.g. rainbow table attack). Setting the cost to minimum almost nullifies the effect of bcrypt (which is not high by modern standards, even on proper cost settings). As we are to use bcrypt according to the task definition, we should at least use a proper cost value.

File: internal/service/users.go [hunk 2]
```diff
@@ -40,6 +45,49 @@ func (u *usersService) UpdateUserInfo(ctx context.Context, id string, input *Upd
...
+	if err := u.repo.Insert(ctx, user); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (u *usersService) Login(ctx context.Context, input *LoginInput) (*core.User, error) {
+	user, err := u.repo.GetByEmail(ctx, input.Email)
+	if err != nil {
```
Comment (zhuravlev-pe) [inline internal/service/users.go]: This approach, combined with what you have in the endpoint implementation, will return different errors in the following cases:

- both email and password are incorrect
- correct email, incorrect password

A normal approach when building login functionality is to reply with the same error in both cases. This is also what we have in the API definition on SwaggerHub: error with status code 400 (bad request) and text "Email or password are incorrect"

**Review by zhuravlev-pe — COMMENTED** [review]
Also, for the next time please use branch name "feature/your_feature_name", to follow Git Flow convention (as per our task)

File: internal/service/users.go (hunk 1 above)
Comment (santhoshMalay) [reply internal/service/users.go:68]: Updated in the latest commit.

File: internal/service/users.go (hunk 2 above)
Comment (santhoshMalay) [reply internal/service/users.go]: Updated in the latest commit.

File: internal/service/users.go [hunk 3]
```diff
@@ -40,6 +45,49 @@ func (u *usersService) UpdateUserInfo(ctx context.Context, id string, input *Upd
...
 
+func (u *usersService) Signup(ctx context.Context, input *SignupUserInput) error {
+	user, err := u.repo.GetByEmail(ctx, input.Email)
+	if err != nil && err != repository.ErrNotFound {
+		return ErrInvalidCredentials
+	}
+
+	if user != nil && user.Email == input.Email {
+		return ErrUserAlreadyExist
+	}
```
Comment (mancerayder515) [inline internal/service/users.go]: `ErrInvalidCredentials` is not on its place, I believe. 

Also, I couldn't think of a case, where an `Email` of `user` and `input` will differ, besides when our repository's `GetByEmail` is broken. Correct me, if i'm wrong.

Comment (zhuravlev-pe) [reply internal/service/users.go]: Yes, you are right. `ErrInvalidCredentials` is supposed to be used in Login(), not here. And check `user.Email == input.Email` is indeed superfluous (but does no harm either)

File: internal/service/users.go [hunk 4]
```diff
@@ -40,6 +45,54 @@ func (u *usersService) UpdateUserInfo(ctx context.Context, id string, input *Upd
...
+	user, err := u.repo.GetByEmail(ctx, input.Email)
+	if err != nil && err != repository.ErrNotFound {
+		return nil, err
+	}
+
+	if err == repository.ErrNotFound {
+		return nil, ErrInvalidCredentials
+	}
+
+	if err = bcrypt.CompareHashAndPassword(user.HashedPassword, []byte(input.Password)); err != nil {
```
Comment (zhuravlev-pe) [inline internal/service/users.go:100]: You are still returning a different error if the password does not match. The whole point of introducing the custom error ErrInvalidCredentials was to return it both in line 87 above and here

**Review by zhuravlev-pe — APPROVED** [review]
Seems to be working, but it definitely needs validation in the signup. This is better be done in a separate PR, using the tools I've provided in my PR (#4 )

File: internal/delivery/http/v1/auth.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+		return
+	}
+	result, err := h.services.Users.Login(ctx.Request.Context(), &input)
+
+	if err != nil {
+		// TODO: discriminate between validation errors, logic errors and internal server errors
+		if err == repository.ErrNotFound {
+			utils.ErrorResponse(ctx, http.StatusNotFound, err)
+			return
+		}
```
Comment (mancerayder515) [inline internal/delivery/http/v1/auth.go]: This clause is never being executed since we don't return `repository.ErrNotFound`.

**Review by mancerayder515 — CHANGES_REQUESTED** [review]

File: internal/delivery/http/v1/auth.go (hunk 5 above)
Comment (santhoshMalay) [reply internal/delivery/http/v1/auth.go]: Removed this. Also, updated the signup implementation as well.
IN both the cases updating the error response with _**StatusBadRequest**_

File: internal/service/users.go (hunk 4 above)
Comment (santhoshMalay) [reply internal/service/users.go:100]: updated in latest commit

**Review by mancerayder515 — APPROVED** [review]

**Pull request merged**